### PR TITLE
iOS Custom Action Parsing

### PR DESCRIPTION
**Pull request by nesalang**
remerged the branch to clean up the history
- has dependency to #2332; should be merged after #2332
todo
- add uwp custom action element

### CODE REVIEW

File: samples/Tests/CustomParsingTestUsingProgressBar.json [hunk 1]
```diff
@@ -15,13 +15,15 @@
...
+					"type": "ProgressBar",
 					"payload": [
 									{
 									"red":187,
 									"green":160,
 									"blue":209,
-									"alpha":1
+									"alpha":1,
+                                    "progress":0.95,
+                                    "thickness":1
```
Comment (paulcam206) [inline samples/Tests/CustomParsingTestUsingProgressBar.json]: spaces?

File: source/android/adaptivecards/src/AdaptiveCardObjectModel.i [hunk 2]
```diff
@@ -101,6 +101,8 @@ struct tm {
 #include "../../../shared/cpp/ObjectModel/Media.h"
 #include "../../../shared/cpp/ObjectModel/ToggleVisibilityAction.h"
 #include "../../../shared/cpp/ObjectModel/ToggleVisibilityTarget.h"
+%include "../../../shared/cpp/ObjectModel/UnknownElement.h"
+%include "../../../shared/cpp/ObjectModel/UnknownActionElement.h"
```
Comment (paulcam206) [inline source/android/adaptivecards/src/AdaptiveCardObjectModel.i]: these should go in the list of `%include` directives below

File: samples/Tests/CustomParsingTestUsingProgressBar.json (hunk 1 above)
Comment (nesalang) [reply samples/Tests/CustomParsingTestUsingProgressBar.json]: thanks

---
In reply to: [251641777](https://github.com/Microsoft/AdaptiveCards/pull/2325#discussion_r251641777) [](ancestors = 251641777)

File: source/android/adaptivecards/src/AdaptiveCardObjectModel.i (hunk 2 above)
Comment (nesalang) [reply source/android/adaptivecards/src/AdaptiveCardObjectModel.i]: thanks

---
In reply to: [251642200](https://github.com/Microsoft/AdaptiveCards/pull/2325#discussion_r251642200) [](ancestors = 251642200)

File: source/ios/AdaptiveCards/ADCIOSVisualizer/ADCIOSVisualizer/CustomActionNewType.mm [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+#import <AdaptiveCards/ACOBaseActionElementPrivate.h>
+#import <AdaptiveCards/SharedAdaptiveCard.h>
+
+@implementation CustomActionNewType
+
+-(instancetype)init
+{
+    self = [super init];
+    return self;
+}
```
Comment (bviglietta) [inline source/ios/AdaptiveCards/ADCIOSVisualizer/ADCIOSVisualizer/CustomActionNewType.mm]: You can leave this out and just let the parent's `init` method be inherited implicitly.

Comment (nesalang) [reply source/ios/AdaptiveCards/ADCIOSVisualizer/ADCIOSVisualizer/CustomActionNewType.mm]: thanks

---
In reply to: [252495022](https://github.com/Microsoft/AdaptiveCards/pull/2325#discussion_r252495022) [](ancestors = 252495022)

File: source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACRRegistration.h [hunk 4]
```diff
@@ -20,16 +20,28 @@
...
-- (void) setCustomElementParser:(NSObject<ACOIBaseCardElementParser> *)customElementParser;
+- (void)setCustomElementParser:(NSObject<ACOIBaseCardElementParser> *)customElementParser key:(NSString *)key;;
 
-- (BOOL) isElementRendererOverriden:(ACRCardElementType)cardElementType;
+- (NSObject<ACOIBaseCardElementParser> *)getCustomElementParser:(NSString *)key;
 
-- (BOOL) isActionRendererOverriden:(NSNumber *)cardElementType;
+- (void)setCustomElementRenderer:(ACRBaseCardElementRenderer *)renderer key:(NSString *)key;
+
+- (BOOL)isElementRendererOverriden:(ACRCardElementType)cardElementType;
```
[the remark this answers is not in the document]
Comment (nesalang) [reply source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACRRegistration.h]: thanks

---
In reply to: [252493940](https://github.com/Microsoft/AdaptiveCards/pull/2325#discussion_r252493940) [](ancestors = 252493940)

**Comment by paulcam206** [comment]
Looks like there's a build failure for iOS (missing reference to `UnknownAction.h`?). Android might just need a rerun...

**Comment by nesalang** [comment]
yes, you are correct! i changed file name.

---
In reply to: [459459597](https://github.com/Microsoft/AdaptiveCards/pull/2325#issuecomment-459459597) [](ancestors = 459459597)

**Pull request merged**